x/interchainquery/keeper/msg_server.go

```diff
@@ -28,105 +28,143 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
-					return nil, fmt.Errorf("unable to verify proof: %s", err)
-				}
-				k.Logger(ctx).Info(fmt.Sprintf("Proof validated! module: %s, queryId %s", types.ModuleName, q.Id))
-
-			} else {
-				// if we got a nil response, verify non inclusion proof.
-				if err := merkleProof.VerifyNonMembership(tmclientstate.ProofSpecs, consensusState.GetRoot(), path); err != nil {
-					return nil, fmt.Errorf("unable to verify proof: %s", err)
-				}
-				k.Logger(ctx).Info(fmt.Sprintf("Non-inclusion Proof validated! module: %s, queryId %s", types.ModuleName, q.Id))
-			}
+	if q.Ttl < curT {
+		errMsg := fmt.Sprintf("[ICQ Resp] aborting query callback due to ttl expiry! ttl is %d, time now %d for query of type %s with id %s, on chain %s",
+			q.Ttl, ctx.BlockHeader().Time.UnixNano(), q.QueryType, q.ChainId, msg.QueryId)
+		k.DeleteQuery(ctx, msg.QueryId)
+		k.Logger(ctx).Error(errMsg)
+		return true, nil
+	}
+	return false, nil
+}
+
+func (k msgServer) SubmitQueryResponse(goCtx context.Context, msg *types.MsgSubmitQueryResponse) (*types.MsgSubmitQueryResponseResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// 1. check if the response has an associated query stored on stride
+	q, found := k.GetQuery(ctx, msg.QueryId)
+	if found {
+		// 2. verify the query's ttl is unexpired
+		exceeded, err := k.HasQueryExceededTtl(ctx, msg, q)
+		if err != nil {
+			return nil, err
+		}
+		if exceeded {
+			return &types.MsgSubmitQueryResponseResponse{}, nil
 		}
 
-		moduleNames := []string{}
-		for moduleName := range k.callbacks {
-			moduleNames = append(moduleNames, moduleName)
+		// 3. verify the response's proof, if one exists
+		err = k.VerifyKeyProof(ctx, msg, q)
+		if err != nil {
+			return nil, err
 		}
-		sort.Strings(moduleNames)
-
-		for _, module := range moduleNames {
-			k.Logger(ctx).Info(fmt.Sprintf("Executing callback for queryId (%s), module (%s)", q.Id, module))
-			moduleCallbackHandler := k.callbacks[module]
-
-			if moduleCallbackHandler.Has(q.CallbackId) {
-				k.Logger(ctx).Info(fmt.Sprintf("ICQ Callback (%s) found for module (%s)", q.CallbackId, module))
-				err := moduleCallbackHandler.Call(ctx, q.CallbackId, msg.Result, q)
-				if err != nil {
-					k.Logger(ctx).Error(fmt.Sprintf("error in ICQ callback, error: %s, msg: %s, result: %v, type: %s, params: %v", err.Error(), msg.QueryId, msg.Result, q.QueryType, q.Request))
-					return nil, err
-				}
-			} else {
-				k.Logger(ctx).Info(fmt.Sprintf("ICQ Callback not found for module (%s)", module))
-			}
+
+		// 4. call the query's associated callback function
+		err = k.FindAndInvokeCallback(ctx, msg, q)
+		if err != nil {
+			return nil, err
 		}
 
+		// 5. delete the query from the store once it's been processed
 		k.DeleteQuery(ctx, msg.QueryId)
 
 	} else {
-		k.Logger(ctx).Info("Ignoring duplicate query")
+		k.Logger(ctx).Info("Ignoring duplicate or nonexistent query")
```

Comment:
- how? `found` will be false, as the query on stride will not be found (will have been already deleted)
- why? because we would like to noop if the query is a duplicate

Reply:
I see - maybe clarify the log to "Ignoring duplicate response..."